internal/cgrouputils/register.go

```diff
@@ -0,0 +1,47 @@
+package cgrouputils
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+)
+
+// Memory cgroup root path
+const (
+	MemRoot string = "/sys/fs/cgroup/memory"
+)
+
+// RegisterMemoryThreshold registers an eventfd(2) and a memory threshold in
+// 'cgroup.event_control' of a cgroup. The eventfd will trigger when
+// memory.usage_in_bytes exceeds the threshold set.
+// https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt Section 9
+func RegisterMemoryThreshold(cgPath string, threshold int64) (*os.File, error) {
+	efd, err := unix.Eventfd(0, unix.EFD_CLOEXEC)
+	if err != nil {
+		return nil, err
+	}
+
+	efdFile := os.NewFile(uintptr(efd), "efd")
+	defer func() {
+		if err != nil {
```

Comment:
I think this won't do what you want unless you name the error return value `err` above, e.g.

`func RegisterMemoryThreshold(cgPath string, threshold int64) (_ *os.File, err error) {`

Reply:
Are you sure? I made a dummy program to test this and it seems to function. err is already defined above

Reply:
This way can work in some cases, but it's not very robust, and simple refactorings can inadvertently cause the clean up code to stop working.

For instance, in the below example, clean up does not occur:

```
package main

import "fmt"

func main()        { fmt.Printf("error: %s\n", test()) }
func works() error { return nil }
func fails() error { return fmt.Errorf("failure") }

func test() error {
	err := works()
	if err != nil {
		return err
	}
	defer func() {
		if err != nil {
			fmt.Println("cleaning up")
		}
	}()
	if err := fails(); err != nil {
		return err
	}
	return nil
}
```

This is because now the scope of the second `err` declaration is in a nested block, so the `defer` doesn't see this value. Instead, it only sees the `err` value in the top-level scope. However, if you use a named return value, it will work correctly.

Reply:
This is what I was testing with which is pretty much this scenario but that makes sense. Lets change it to be safe. 
```go
func errorTest() (string, error) {
	err := fmt.Errorf("oh no an error")
	defer func(){
		if err != nil {
			fmt.Println("oh no again")
		}
	}()
	if err != nil {
		return "", err	
	}
	return "success", nil	
}```